Studies/source/PuritySignalStudy.cxx

```diff
@@ -0,0 +1,306 @@
+    {
+        std::vector<analysis::jet_ordering::JetInfo<LorentzVector>> jet_info_vector;
+        BTagger bTagger(Period::Run2017,jet_ordering);
+
+        for(size_t jet_index = 0; jet_index < event.jets_p4.size(); ++jet_index)
+            jet_info_vector.emplace_back(static_cast<LorentzVectorE>(event.jets_p4.at(jet_index)), jet_index, bTagger.BTag(event,jet_index));
+
+
+        return jet_ordering::OrderJets(jet_info_vector, false, 20, 2.1);
+    }
+
+    std::vector<size_t> orderTaus(const Event& event, TauIdDiscriminator tau_id_discriminator, bool is_descending_order = true)
+    {
+       std::vector<float> raw_values;
+       for(size_t lep_index = 0; lep_index < event.lep_p4.size(); ++lep_index){
+           if (static_cast<LegType>(event.lep_type.at(lep_index)) == analysis::LegType::tau ) {
+               auto lepton = std::make_shared<ntuple::TupleLepton>(event, lep_index);
+               auto raw = static_cast<float>(lepton->GetRawValue(tau_id_discriminator));
+               raw_values.push_back(raw);
+           }
+           else
+                raw_values.push_back(0);
+       }
+
+
+       std::vector<size_t> sorted_had_taus;
+       std::vector<size_t> sorted = sort_indexes(raw_values, is_descending_order);
+
+       for (const auto& index : sorted){
+               if(isHadronicTau(static_cast<LegType>(event.lep_type.at(index))))
+                   sorted_had_taus.push_back((index));
+       }
+       return sorted_had_taus;
+    }
+
+    std::vector<size_t> sort_indexes(const std::vector<float> &v, bool is_descending_order = true) {
+      // initialize original index locations
+      std::vector<size_t> idx(v.size());
+      iota(idx.begin(), idx.end(), 0);
+
+      // sort indexes based on comparing values in v
+      sort(idx.begin(), idx.end(),
+           [&v,is_descending_order](size_t i1, size_t i2) {
+               if(!is_descending_order)
+                  return v[i1] < v[i2];
+               else
+                  return v[i1] > v[i2];
+      });
+
+      return idx;
+    }
+
+    std::vector<size_t> intersection(std::vector<size_t> &v1, std::vector<size_t> &v2)
+    {
+        std::vector<size_t> v3;
+
+        std::sort(v1.begin(), v1.end());
+        std::sort(v2.begin(), v2.end());
+
+        std::set_intersection(v1.begin(),v1.end(),
+                              v2.begin(),v2.end(),
+                              back_inserter(v3));
+        return v3;
+    }
+
+    void printStats(const TH1D& histo, int binIndex, std::string tag)
+    {
+        double histoContent = histo.GetEntries();
+        double binContent = histo.GetBinContent(binIndex);
+        double ratio = binContent / histoContent;
+        std::cout << "using bin #" << binIndex << ", rate of " << tag << " : " << binContent << "/" <<  histoContent
+                  << " = " << ratio << std::endl;
+    }
+
+    static bool isHadronicTau(const LegType& leg_type)
```

Comment:
do comparison `leg_type == LegType::tau` directly in the code, don't need to define a function for that.